rclcpp_components/cmake/rclcpp_components_add_library_with_nodes.cmake

```diff
@@ -0,0 +1,39 @@
+# Copyright 2019 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Create a target which builds a shared library and also creates a target
+# which builds an executable
+# usage: rclcpp_components_add_node(
+#              <target> [EXCLUDE_FROM_ALL] <source1> [source2 ...])
+
+# :param EXCLUDE_FROM_ALL: exclude the target from default build target
+# :type EXCLUDE_FROM_ALL: boolean
+# :param target: the name of the shared library target
+# :type target: string
+# :param sourceN: the list of source files for executable and shared library
+# :type sourceN: list of strings
+#
+function(rclcpp_components_add_library_with_nodes target)
+  cmake_parse_arguments(
+    ARGS
+    "EXCLUDE_FROM_ALL"
+    ""
+    ""
+    ${ARGN})
+  set(sourceN ${ARGS_UNPARSED_ARGUMENTS})
+  add_library(${target} SHARED ${sourceN})
+  string(TOUPPER ${PROJECT_NAME} PROJECT_NAME_UPPER)
+  target_compile_definitions(${target} PRIVATE "${PROJECT_NAME_UPPER}_BUILDING_DLL")
+  set_target_properties(${target} PROPERTIES EXCLUDE_FROM_ALL ${ARGS_EXCLUDE_FROM_ALL})
+endfunction()
```

Comment:
I'm wondering now if you even need this function? It is essentially just calling `add_library`, right?

Reply:
Is the `_BUILDING_DLL` really needed?

Reply:
I was keeping `_BUILDING_DLL` because of the Windows issues I was having but not sure if that was even doing anything. Other than that, I don't think that functions needed then

Reply:
Let's try it without? I think that definition is something the user can do themselves. It doesn't work anyways unless they create a corresponding header file, or use CMake to generate one.

Reply:
That way it just becomes `add_library` and `rclcpp_components_register_node`.